### PR TITLE
Weighted partition

### PR DESCRIPTION
**Pull request by peterlimg**
This  implements the partition weights for challenge-ready blobber partitions with metadata weights.

When selecting a blobber to challenge, a random value in the range [0, total weight of all partitions] is generated.
The partition is then selected based on the random value, and a blobber is picked from that partition.
The implementation details of this selection process can be found in the `pick()` method.

In addition to the `pick()` method, the core function of this partition weights is to keep the blobber weights and partition weights in sync, especially when a blobber weight is added, updated or removed.
## Changes
- ~Add blobber weights wrapper that manage the blobber weights partitions and the node to track partition weights~
- Add random select base on random value from [0, total weights of all partitions]. 
- Add methods to migrates from blobber challenge ready partitions to the blobber weights when hard fork happens
- Add method to add/update/remove blobber to sync the weights between blobbers partitions and partition weights
- Add code to switch the selection logic with activator
## TODOs:
- [ ] Test on network
- [x] Call the add/update/remove methods to sync weights when blobber is added/updated/removed from the challenge ready partitions.
- [x] Enable with hardfork activator

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by peterlimg** [comment]
Replace the blobber weights partition with the existing challenge ready blobber partitions to avoid duplicate partitions.

File: code/go/0chain.net/smartcontract/partitions/partitions.go [hunk 1]
```diff
@@ -178,27 +198,55 @@ func (p *Partitions) pack(state state.StateContextI) error {
...
-		return err
+		return -1, err
 	}
 
 	p.loadLocations(loc)
+	return loc, nil
+}
+
+// ForEachPart iterates all items in specific partition,
+// break whenever the callback function returns false
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/partitions/partitions.go:226]: breaks when returns true

File: code/go/0chain.net/smartcontract/storagesc/partitions_weights.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+						// find the blobber, break and return
+						stop = true
+					}
+					return
+				}); err != nil {
+				return "", err
+			}
+
+			if blobberID == "" {
+				return "", fmt.Errorf("could not pic a blobber, blobber weights may not synced")
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/storagesc/partitions_weights.go:76]: typo pick

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**